models/experimental/grok/tt/grok_attention.py

```diff
@@ -0,0 +1,347 @@
+            cache_file_name=cache_name(f"wqkv_multidevice_4d"),
+        )
+
+        self.wo = ttnn.as_tensor(
+            torch.transpose(
+                self.state_dict[wo_str],
+                -2,
+                -1,
+            )
+            .unsqueeze(0)
+            .unsqueeze(0),
+            device=self.device_mesh,
+            mesh_mapper=ReplicateTensorToMesh(self.device_mesh),
+            dtype=self.dtype,
+            memory_config=self.model_config["ATTN_WEIGHTS_MEMCFG"],
+            layout=self.model_config["ATTN_W_LAYOUT_TILE"],
+            cache_file_name=cache_name(f"wo_multidevice4d_H"),
+        )
+
+        cache_k = torch.zeros(
+            (
+                self.n_kv_heads,
+                self.max_batch_size,
+                self.max_seq_len,
+                self.head_dim,
+            )
+        )
+        cache_v = torch.zeros(
+            (
+                self.n_kv_heads,
+                self.max_batch_size,
+                self.max_seq_len,
+                self.head_dim,
+            )
+        )
+        layer_past = [cache_k, cache_v]
+        self.layer_past = [
+            ttnn.as_tensor(
+                lp,
+                device=self.device_mesh,
+                mesh_mapper=ShardTensorToMesh(self.device_mesh, dim=0),
+                dtype=ttnn.bfloat8_b,
+                layout=self.model_config["ATTN_W_LAYOUT_TILE"],
+                memory_config=self.model_config["ATTN_CACHE_WEIGHTS_MEMCFG"],
+                cache_file_name=cache_name(f"empty_attn_cache_{cache_k.shape}"),
+            )
+            for lp in layer_past
+        ]
+
+        self.compute_kernel = self.model_args.get_compute_kernel_config()
+        self.compute_kernel_attn = self.model_args.get_compute_kernel_attn_config()
+
+        self.core_grid = self.model_args.max_grid_size
+        self.core_grid_attention = self.model_args.core_grid_attention
+
+        # Will be filled during the initial warmup run
+        self.q_mem_config = None
+        self.k_mem_config = None
+
+    def forward(
+        self,
+        xs,
+        current_pos,
+        attn_masks,
+        rot_mats,
+    ):
+        """
+        x: (seq_len, 1, batch, hidden_dim)
+        current_pos: the length of the KV cache. Same as current token's index.
+        attn_masks: (seq_len, batch, n_heads, cache_len+seq_len)
+        rot_mats: list of rotation matrices for each device
+
+        Tensors are postfixed with 4 characters that represent their 4-D shape:
+        B : batch_size (32)
+        H : dim (6144)
+        D : head_dim (128)
+        P : padded_layer_past_len
+        """
+        padded_layer_past_len = min(nearest_32(current_pos + 1), self.max_seq_len)
+
+        x_11BH = xs
+        wo = self.wo
+        layer_past = self.layer_past
+        rot_mat = rot_mats[current_pos]
+        attn_mask_1B4P = attn_masks
+        ###
+        # QKV matmuls
+        ###
+
+        xqkv_fused = ttnn.experimental.operations.primary.matmul(
```

Comment:
this can be just `ttnn.matmul`

Reply:
Thanks, rebased and updated all of these!